OSDOCS#11845: Multiarch Tuning Operator TP to GA upgrade (#80786)
openshift/openshift-docs: post_installation_configuration/configuring-multi-arch-compute-machines/multiarch-tuning-operator.adoc

toc::[]

The Multiarch Tuning Operator enhances the operational experience within multi-architecture clusters, and single-architecture clusters that are migrating to a multi-architecture compute configuration.

> Architecture-aware workload scheduling allows the default {product-title} pod scheduler to place pods onto nodes that match the architecture of the pod's images. 

Reviewer: ```suggestion
Architecture-aware workload scheduling allows the default scheduler to place pods onto nodes that match the architecture of the pod's images. 
```

No pod I think... but also do we need to specify {product-title}? I'm not sure how the kube scheduler is referred in other pages of the doc.

Author: @aleskandro 

There are instances of "pod scheduler" and "{product-title} pod scheduler" in the docs:
Link - https://docs.openshift.com/container-platform/4.15/nodes/scheduling/nodes-scheduler-about.html

![Screenshot from 2024-08-29 12-26-11](https://github.com/user-attachments/assets/72646e40-9200-4c8c-8b3f-2eb5967d9f47)

> The default pod scheduler does not natively consider the architecture of a pod's images when determining the placement of new pods onto nodes. This Operator implements the `clusterpodplacementconfigs` custom resource (CR) to support architecture-aware workload scheduling.

Reviewer: ```suggestion
The default scheduler does not natively consider the architecture of a pod's images when determining the placement of new pods onto nodes. This Operator implements the `clusterpodplacementconfigs` custom resource (CR) to support architecture-aware workload scheduling.
```

Author: Updated as per the suggestion.

toc::[]

The Multiarch Tuning Operator enhances the operational experience within multi-architecture clusters, and single-architecture clusters that are migrating to a multi-architecture compute configuration.

> Scheduling pods on nodes that support the architecture of the pod's images is called architecture-aware workload scheduling. This Operator implements the `clusterpodplacementconfigs` custom resource (CR) to support architecture-aware workload scheduling.

Reviewer: ```suggestion
Architecture-aware workload scheduling allows a cluster to assign pods to nodes that match the architecture of the pod's images. The Kubernetes scheduler does not natively consider the architecture of a pod's images when selecting nodes for placement. This Operator implements the clusterpodplacementconfigs custom resource (CR) to enable architecture-aware scheduling.
```

What about this ^?

Author: It looks good! Updated as per the suggestion.

By default, the scheduler does not consider the architecture of a pod's container images when determining the placement of new pods onto nodes.

To enable architecture-aware workload scheduling, you must create the `ClusterPodPlacementConfig` object. When you create the `ClusterPodPlacementConfig` object, the Multiarch Tuning Operator deploys the necessary operands to support architecture-aware workload scheduling.

When a pod is created, the operands perform the following actions: 

. Add the `multiarch.openshift.io/scheduling-gate` scheduling gate that prevents the scheduling of the pod.
. Compute a scheduling predicate that includes the supported architecture values for the `kubernetes.io/arch` label. 
. Integrate the scheduling predicate as a `nodeAffinity` requirement in the pod specification.
. Remove the scheduling gate from the pod.
 
[IMPORTANT]
====
Note the following operand behaviors:

* If the `nodeSelector` field is already configured with the `kubernetes.io/arch` label for a workload, the operand does not update the `nodeAffinity` field for that workload.

> * If the `nodeSelector` field is not configured with the `kubernetes.io/arch` label for a workload, the operand updates the `nodeAffinity` field for that workload. However, in that `nodeAffinity` field, the operand updates only the node selector terms that are not configured with the `kubernetes.io/arch` label.

Reviewer 2: @aleskandro @sr1kar99 do we need to point `nodeName` field here?  becasue if `nodeName` is configured, the operand will not gate pod too.

Reviewer: you're right @lwan-wanglin we can add this too:

> If the `nodeName` of the pod is already set, the scheduling is skipped and the operator does not process the pod.

Author: Updated as per the suggestion.

Reviewer 2: thanks, looks good to me now

toc::[]

The Multiarch Tuning Operator optimizes workload management within multi-architecture clusters and in single-architecture clusters transitioning to multi-architecture environments.

> Architecture-aware workload scheduling allows the scheduler to place pods onto nodes that match the architecture of the pod's images. 

Reviewer 3: Avoid possessive for pod: https://www.ibm.com/docs/en/ibm-style?topic=grammar-possessives
```suggestion
Architecture-aware workload scheduling allows the scheduler to place pods onto nodes that match the architecture of the pod images. 
```

Author: Updated as per the suggestion.

When a pod is created, the operands perform the following actions: 

. Add the `multiarch.openshift.io/scheduling-gate` scheduling gate that prevents the scheduling of the pod.
. Compute a scheduling predicate that includes the supported architecture values for the `kubernetes.io/arch` label. 
. Integrate the scheduling predicate as a `nodeAffinity` requirement in the pod specification.
. Remove the scheduling gate from the pod.
 
[IMPORTANT]
====
Note the following operand behaviors:

* If the `nodeSelector` field is already configured with the `kubernetes.io/arch` label for a workload, the operand does not update the `nodeAffinity` field for that workload.

* If the `nodeSelector` field is not configured with the `kubernetes.io/arch` label for a workload, the operand updates the `nodeAffinity` field for that workload. However, in that `nodeAffinity` field, the operand updates only the node selector terms that are not configured with the `kubernetes.io/arch` label.

> * If the `nodeName` of the pod is already set, the Multiarch Tuning Operator does not process the pod.

Reviewer 2: Can we change it align with above description? like below
```suggestion
* If the `nodeName` field is already set, the Multiarch Tuning Operator does not process the pod.
```

Author: Yes, this makes sense! Good catch :) 
updated as per the suggestion. @lwan-wanglin 
Thanks!

[IMPORTANT]
====

> If you are using node affinity and node selectors in the same pod configuration, note the following operand behaviors:

Reviewer: ```suggestion
```

It doesn't matter if users are using one, the other or both.

Author: How about updating it to just "Note the following operand behaviors:

* If the `nodeSelector` field is already configured with the `kubernetes.io/arch` label for a workload, the operand does not update the `nodeAffinity` field for that workload.

> * If the `nodeSelector` field is not configured with the `kubernetes.io/arch` label for a workload, the operand updates the `nodeAffinity` field for that workload. However, in that `nodeAffinity` field, the operand updates only the `nodeSelectorTerms` fields or `matchExpressions` fields that are not configured with the `kubernetes.io/arch` label.

Reviewer: ```suggestion
* If the `nodeSelector` field is not configured with the `kubernetes.io/arch` label for a workload, the operand updates the `nodeAffinity` field for that workload. However, in that `nodeAffinity` field, the operand updates only the node selector terms that are not configured with the `kubernetes.io/arch` label.
```

`nodeSelectorTerms` is not a visible field to the users: it's a "generic name" that can represent a list of matchFields, matchExpressions, or both. It's a defined API but since it's not visible to the users, should we avoid writing it in camelCase?

Author: @aleskandro 

I noticed that `nodeSelectorTerms` is specified similar to other fields in the pod specification.

![Screenshot from 2024-08-27 23-28-14](https://github.com/user-attachments/assets/55a3f7d5-46c8-491b-96c9-ae440cf35794)
https://docs.openshift.com/container-platform/4.16/nodes/scheduling/nodes-scheduler-node-affinity.html

However, I agree with you on using just "nodeSelctorTerms" to refer to all other fields that it can include (such as matchFields, matchExpressions).

Since there are instances in the documentation that use the term `nodeSelectorTerms`, can we keep it this way:

"However, in that `nodeAffinity` field, the operand updates only the `nodeSelectorTerms` that is not configured with the `kubernetes.io/arch` label."

We can also have multiple `nodeSelectorTerms`, in that case we should rather say "`nodeSelectorTerms` fields that are not configured". Can we refer to `nodeSelectorTerms` as a field?
